### ISSUE TITLE
Gradual strategy to disallow usage of internal APIs

### ISSUE THREAD

**Issue by melix** [gradle/gradle#13501]
## context

May (many) plugins and build scripts use internal APIs, just because they are visible at compile time or run time. This causes a number of issues when we want to evolve code and it gives a wrong impression of instability to users: we make sure we don't break any public APIs, but not internal APIs: for this we "only" rely on smoke testing of popular plugins.

We should slowly move out of this situation by applying a gradual strategy. We can do one of more of this ideas, in any particular order: 

a. introduce an `@InternalApi` annotation which could be applied to new internal packages or types in existing packages, making them effectively invisible in the `gradleApi()`
b. Change the gradleApi() dependency so that it marks all internal things as @deprecated and maybe also filters some stuff out. This would work right now for everything other than Groovy DSL scripts and reflective access. Maybe that’s a good enough start, but we could also change the Groovy DSL scripts to compile against the Gradle API as well, and maybe finally change gradleApi() to provide API and runtime variants (we can generate the API much faster, and it’s also much smaller and so could be bundled in the distro). Another downside is that we can’t give great error messages or guidance about what to do.
c. Apply warnings or more filters to the Gradle runtime classloader, so that when a compiled script/plugin/whatever attempts to reference an internal type, we spit out a warning. The advantage is that the plugin author cannot ignore these, but the downside is that we don’t easily know where the usage comes from. Another downside is that the warning happens only when the type is used at runtime, which may not happen at plugin development time when the plugin does not have great coverage.
d. Apply this when we instrument the build logic. We now have some infrastructure to do all kinds of crazy inspections of build logic prior to attempting to load it. We could complain about references to internal types, and also pin point where the reference occurs. This has the upside of working for all scripts and plugins, but for now will miss reflective access. This is also a big improvement, though. It also will apply if the plugin has any testkit tests.
e. Scan the plugins published in the Gradle portal and flag them when they use internal APIs, so that we can show to the users that it's potentially breaking
f. Scan plugin code before publishing to the portal and add an opt-in flag if you use internal APIs, as a "I understand the risks"

**Comment by eskatos** [comment]
Maybe what Kotlin is doing with `@RequiresOptIn` and `@OptIn` could be of inspiration
https://kotlinlang.org/docs/reference/opt-in-requirements.html

**Comment by big-guy** [comment]
What about something like...
- Introduce a `Dependency gradleApi(GradleApi access)`
- `enum GradleApi { PUBLIC_ONLY, PUBLIC_AND_INTERNAL; }`
- `gradleApi()` is assumed to be `PUBLIC_AND_INTERNAL`. 
- When access is `PUBLIC_ONLY`, we always strip all internal packages
- At some point, we emit a deprecation warning that your plugin build is using `PUBLIC_AND_INTERNAL` and we suggest that people switch to `PUBLIC_ONLY`.
- A little later, we swap the default behavior of `gradleApi()` to be `PUBLIC_ONLY`.

We could figure out some way to publish these details too.

**Comment by melix** [comment]
Wouldn't this require generating 2 distinct API jars?

**Comment by adammurdoch** [comment]
I'm not sure we necessarily need a temporary opt in. It might be simpler to just to use a longer deprecation cycle.

**Comment by lacasseio** [comment]
Although preventing plugin authors from accessing the internal API is a great idea, the reality is plugins are using internal API Gradle lack useful public APIs. This data is getting a bit old (about 2 world meetings ago), there were [1548 plugins using internal APIs in the plugin portal](https://gradleplugins.org/#violations). Even Nokee uses internal API and when I try to avoid some internal API the configuration cache report is filled with errors.

In the [Gradle Plugin Toolbox](https://github.com/gradle-plugins/toolbox), I was going to offer something more along the lines of what Sterling said: `gradlePublicApi(version)`/`gradlePublicAndInternalApi(version)`.

**Comment by melix** [comment]
What needs to be fixed is having public APIs for what plugin authors need. Using internal APIs will not help. It makes things more complicated for us, and for users in the end. I am -1 on opt-in for use of internal APIs, as it's effectively encouraging use of them.

**Comment by lacasseio** [comment]
I agree, my point was there is a more substantial investment at the moment to provide all the alternative public API. Either an opt-in or the deprecation cycle will need to be long enough to allow users to migrate.

Another approach suggested some time ago was to generate an internal API usage report to help users migrate away from internal API and use the correct alternative which could be a public API or a third-party library.
